### ISSUE TITLE
Implement HaTS survey triggers for users who answers "Other" to any User Input question 

### ISSUE THREAD

**Issue by kuasha420** [google/site-kit-wp#6180]
## Feature Description

Since we're removing the text field next to the "Other" in all the User Input questions in #6181, we should implement surveys for users who answer other to get more information on their choice. This can be a single survey or multiple surveys, whichever is best UX. 

Related Asana Task: https://app.asana.com/0/1201538818998708/1203393984679795

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* A survey should be triggered for users who answer "Other" to at least one of the User Input questions. 
* This should only be triggered once user completes answering the questions.
    * This should also trigger after user edits user input question and at least one of the answers changes to "Other". 
* For sites with multiple admins, the survey will only appear to the admin who answered the question. This is regardless whether the question is user or site specific.
* Since there are three user input questions and user can answer Other to any combination of them, one of the following trigger IDs should be used based on the answer. 
    * `userInput_answered_other__purpose`
    * `userInput_answered_other__postFrequency`
    * `userInput_answered_other__goals`
    * `userInput_answered_other__purpose_postFrequency`
    * `userInput_answered_other__purpose_goals`
    * `userInput_answered_other__postFrequency_goals`
    * `userInput_answered_other__purpose_postFrequency_goals`
* The the survey returned from the Service will be queued and shown to the user as part of #6306. 

## Implementation Brief

* Create a new action `maybeTriggerUserInputSurvey` action in  `CORE_USER` store. 
    * Get the current user input settings using the `getUserInputSettings` selector.
    * If there is at least one submitted answer contains `other`-
        * Construct one of the survey `triggerID` mentioned in the AC.
        * Using the `triggerSurvey` action from `CORE_USER` trigger the survey for the user using the aforementioned `triggerID`.
* Update the `saveUserInputSettings` action of `CORE_USER`:
    * If there was no error, before setting `isSaving` to false, `yield` the `maybeTriggerUserInputSurvey` action. 
* In addition to triggering the survey, we need to also make sure any existing outdated surveys in the `Survey_Queue` is dequeued. For that: 
    * In `Google\Site_Kit\Core\User_Input\REST_User_Input_Controller` class:
        * Add a protected `$survey_queue` instance variable which will be received and assigned in the constructor.
        * In `POST:core/user/data/user-input-settings` rest route:
            * Before returning the return value of `rest_ensure_response` check if it's a `WP_REST_Response`.
            * Dequeue `user_input_answered_other_survey` survey using the `$survey_queue->dequeue( survey_id )`. 
                * (Note: This is the preliminary `survey_id` for all `userInput` "other" surveys ) 
    * In `Google\Site_Kit\Core\User_Input\User_Input` class:
        * In the constructor, pass a new `Survey_Queue` instance while creating the `REST_User_Input_Controller` class instance. 

### Test Coverage

* Add tests for `maybeTriggerUserInputSurvey` action.
* Add test coverage for `saveUserInputSettings` action.
* Add test coverage for the above PHP changes.

## QA Brief

- Set up a new site, install, and connect Site KIt.
- Using the tester plugin, enable the `userInput` feature flag.
- Go to the `/wp-admin/admin.php?page=googlesitekit-user-input` screen. Answer the user input questions and answer "Other" for one or more questions.
- In the User Input preview screen, open your browser developer tools and go to the "Network" tab. Ensure that network log recording is on so that network requests persist even though you get navigated to a different screen.
- Click "Save".
- Ensure that you see a request to the `/wp-json/google-site-kit/v1/core/user/data/survey-trigger?_locale=user` endpoint that has `{"data":{"triggerID":"userInput_answered_other__[TRIGGER_ID]"}}` in payload.
- The `TRIGGER_ID` placeholder above should include slug of questions that you have answered "Other" for, according to the ACs.
- Answer the questions again but do not choose "Other" this time. Verify that the above network request is not made in this case.

## Changelog entry

* Implement HaTS survey triggers for users who answers "Other" to any User Input question

**Comment by eugene-manuilov** [comment]
Some context from the original Asana ticket:

```
Eugene Manuilov:
@Arafat Zahan, since all answers are stored on the WordPress side, 
how can we know that a user has selected the "Other" answer?

Evan Mattson:
@Eugene Manuilov the plan is to dispatch a survey trigger from the plugin 
side for these users. The idea was to send the question IDs for survey 
questions the user answered "other" for in the trigger but this would require 
somehow including this in the trigger ID or updating the endpoint + trigger
mechanism to allow for passing some arbitrary payload data along with the
trigger request. Does that make sense? cc: @Felix Arntz

Eugene Manuilov:
I would prefer if the plugin sends a request to an endpoint on the service side
to notify about saving the user input settings (maybe for the first time 🤔) to
use this information to determine whether we need to show the survey for
the user.

Evan Mattson:
> I would prefer if the plugin sends a request to an endpoint on the service
> side [...] to determine whether we need to show the survey for the user.

That's kind of what a survey trigger is, right? When we send the trigger and
what data we can send along with it should provide all the information the
service needs to determine that, correct?

We could trigger the survey on user input completion, but we should probably
also only ask the user for input regarding the "other" response for answers which
they answered themselves (as site-specific answers could have been answered
by another user of course).
```

**Comment by eugene-manuilov** [comment]
> When we send the trigger and what data we can send along with it should provide all the information the service needs to determine that, correct?

@aaemnnosttv my main concern about this approach is that it makes the survey triggering logic rely on two sources of truth: the site state stored on the service side and the payload sent with the trigger request. In my opinion this is worse than what we have now. Plus, in the future if we have multiple features that also need to trigger surveys based on different conditions, this approach can lead to a gigantic payload that contains a lot data.

I prefer to use another approach where we somehow let the service know that a site has "other" options in the key metrics and then the service will make a decision whether to show this HaTS survey based on the information saved on the service side. For example, we can send the information that the site has "other" key metrics when we sync the site fields.

@adamsilverstein @felixarntz what do you think?

**Comment by aaemnnosttv** [comment]
@eugene-manuilov I'm not sure I follow your concern. I'm suggesting we send a small amount of extra data along with the trigger (question IDs), and then the service can determine if it should respond with a survey or not, and the steps of the response would be based on the given question IDs. Is it perhaps not as straightforward as I am imagining?

> I prefer to use another approach where we somehow let the service know that a site has "other" options in the key metrics and then the service will make a decision whether to show this HaTS survey based on the information saved on the service side. For example, we can send the information that the site has "other" key metrics when we sync the site fields.

The problem I see with this approach is that it would treat the "other" choice state as site-wide, where at least part of it is user-specific.

Perhaps there is more of a service-specific architectural nuance to this but I would think having the ability to include a small payload with the trigger request would be useful here and in the future.

**Comment by eugene-manuilov** [comment]
> Is it perhaps not as straightforward as I am imagining?

@aaemnnosttv, it is very straightforward.

> I'm not sure I follow your concern.

I think that the decision which survey to show for a user should be based on the information that the service has collected about the site and the user. Not on the information that has been passed with the request, because it can be easily spoiled by a 3rd party plugin that just modifies the request URL using WP hooks.

> I'm suggesting we send a small amount of extra data along with the trigger (question IDs)...

Yes, today it is a small amount of extra data. Tomorrow, if we implement a new feature or change something in existing modules, we will have a temptation to send a little bit more data to cater for new needs. In a month, we will want to send more data again, because we have an urgent request and we can't wait: let's use a straightforward approach and send :allthethings: with the trigger. 😄 

> > I prefer to use another approach where we somehow let the service know that a site has "other" options in the key metrics and then the service will make a decision whether to show this HaTS survey based on the information saved on the service side. For example, we can send the information that the site has "other" key metrics when we sync the site fields.
> 
> The problem I see with this approach is that it would treat the "other" choice state as site-wide, where at least part of it is user-specific.

We can make it user specific by using hashed user identifiers.

**Comment by kuasha420** [comment]
@felixarntz sounds good to me. Can we create a issue for the enhancement of survey trigger mechanism so we can make this issue blocked by that one, to make things super clear to the wider team? Cheers!

**Comment by felixarntz** [comment]
@kuasha420 Good point, done!

**Comment by bethanylang** [comment]
@kuasha420 Following up here, thanks!

**Comment by kuasha420** [comment]
@bethanylang This one is still blocked by 6306, however we had a call just before the break and made some decisions and progress on it. There's a doc about it in the open slack channel and @aaemnnosttv is working on adding preliminary ACs to #6306. Once it moves forward, we will revisit this issue. Cheers!

**Comment by tofumatt** [comment]
ACs look good here and I think cover the issue, so moving to IB 👍🏻

**Comment by aaemnnosttv** [comment]
Thanks @kuasha420 – a few thoughts for you

* Would it maybe be easier to trigger the survey at the time that the user input settings are being saved? If not on the backend, we could do this on the dashboard conditionally based on the `notification=user_input_success`
	* If on the client, `UserInputSuccessBannerNotification` would make more sense than on the dashboard I think
	* Regarding the `ttl` - this might not be relevant since the questions aren't going to be answered by the same user multiple times in a row - I think the above condition might handle this part
* One thing we might want to consider is the potential for multiple surveys related to the above trigger IDs being queued simultaneously (i.e. a user answered once, triggered one, queued that survey and then answered again in a way that invoked a different trigger – would we queue another similar survey?). There's of course a service element to this but one we should foresee and account for.
* Test coverage is empty

**Comment by kuasha420** [comment]
@aaemnnosttv good observations. 

* Adding the Trigger in `UserInputSuccessBannerNotification` won't fully satisfy the AC because we will also have to re-trigger the surveys when Settings are upgraded. We can possibly enhance the `saveUserInputSettings` action to do the triggering as part of it, or we can create a separate action which can be triggered inside the `saveUserInputSettings` or on `UserInputPreviewGroup` (for settings) and `UserInputQuestionnaire` (for setup). I've went ahead and updated the IB to use the first approach, but let me know if you prefer any other solution. ie. serverside, I'd prefer to keep this client site to keep things simple.
* We should queue the latest survey and discard the previous ones service side, if possible. 
* Added test coverage.

**Comment by aaemnnosttv** [comment]
Thanks @kuasha420 

> * We can possibly enhance the `saveUserInputSettings` action to do the triggering as part of it, or we can create a separate action which can be triggered inside the `saveUserInputSettings`

SGTM 👍 This action makes sense as the place to invoke it but is a bit busy as it is, so let's handle the triggering in another action. 

> * We should queue the latest survey and discard the previous ones service side, if possible.

The queue is local so we wouldn't need to make changes on the service. Instead I think we simply need to dequeue all the related "other" trigger IDs than the one we triggered. We don't want to dequeue the same trigger ID if it was already queued (i.e. no change). In order to avoid adding another endpoint to prune other surveys, or handle this trigger as a special case, I think we can a listener on the survey queue meta value – if it changes (only need to consider updates, not adding/deleting), we'd want to get the last user input "other" survey, and then dequeue the other variants. Does that make sense?

**Comment by kuasha420** [comment]
@aaemnnosttv Thank you for the clarification about the queue, which I mixed up with the trigger. I've added a point about it in the IB, however, we need to confirm on the relationship between `triggerID` and `surveyID` as discussed in our call. I'll leave this one with you for now to arrange a sync with @eugene-manuilov and sort it out. 

PS: Bumping estimate for the new point and test coverage.

**Comment by aaemnnosttv** [comment]
> * In `Google\Site_Kit\Core\User_Surveys\User_Surveys`

The survey ID for all of these will be the same and the queue only allows unique survey IDs to be enqueued so the way the hook is described here wouldn't work because enqueuing the same survey more than once wouldn't invoke the filter because the queue wouldn't attempt to update it.

@eugene-manuilov I feel like the simplest solution here would be to allow a survey to be updated/replaced in the queue when it already exists.

This could 

1. be the new default behavior (because updating the a survey with the same payload would be no change, and thus no update) or 
2. we could update the `survey-trigger` endpoint with an optional `update` parameter which would call a new method on the queue which would enqueue/update depending on whether or not the survey was present. This would be a more explicit use that we'd only need here but could be useful in the future

This would also be useful in that we wouldn't need to have special behavior for any specific survey. WDYT?

**Comment by kuasha420** [comment]
@aaemnnosttv Thank you for the review but this was the previous iteration of the IB before we had the call. As par our slack discussion, I've updated the IB to unconditionally dequeue the `userInput_answered_other` survey from the `Survey_Queue` after User Input Settings is saved. **Any potential survey will get triggered afterwards from client-side, so as long as service keeps returning the (incomplete) survey, this approach should be perfectly fine** and also make sure any pending survey is dequeued when the new answer doesn't contain other. 

Please have a look. I'll also keep @eugene-manuilov assigned here so he can confirm the detail about the service related detail I mentioned above. 

Thank you both for all your help, I feel we're very close to finalizing this!

Cheers.

**Comment by aaemnnosttv** [comment]
Thanks @kuasha420 – I tweaked the name of the survey id but otherwise this looks good to me 👍 

IB ✅

**Comment by mohitwp** [comment]
### QA Update ⚠️ 

@nfmohit 

- Tested on dev environment with ga4Reporting and user input feature flag enabled.
- I followed QAB but when I clicked on 'Save' button then no request to` /wp-json/google-site-kit/v1/core/user/data/survey-trigger?_locale=user` is made. Instead, it shows survey timeout when user clicks on 'Save' button.


https://github.com/google/site-kit-wp/assets/94359491/bfbd0a6d-b65a-4fa1-b7f0-bbae87fad6df

**Comment by mohitwp** [comment]
### QA Update ❌ 

Moving to execution as it will need follow-up PR.

**Comment by aaemnnosttv** [comment]
Back to you for another pass @mohitwp !

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Tested on main branch.
- Verified that a request to the `/wp-json/google-site-kit/v1/core/user/data/survey-trigger?_locale=user `endpoint that has {"data":{"triggerID":"userInput_answered_other__[TRIGGER_ID]"}} in payload is made when user select 1 or more time other as a User input survey questions.
- Verified for scenarios when user selects 'OTHER' options as a  ans for 1 question or 2 questions or for all 3 questions.
- Verified that trigger id is getting update as per the answer selection.
- Also, verified for the case when user don't select 'OTHER' as a ans for any question. In this case request to the `/wp-json/google-site-kit/v1/core/user/data/survey-trigger?_locale=user `endpoint is not getting generate.
- Also, verified the same cases when user edit answers.

<details>


When user selects other option

https://github.com/google/site-kit-wp/assets/94359491/705137bd-c3f8-40aa-99d0-5e4991d3b293


https://github.com/google/site-kit-wp/assets/94359491/27678aae-7d50-4567-8c27-ec37fbe4fe01


https://github.com/google/site-kit-wp/assets/94359491/6ba34fd3-8c86-4377-8b7a-331d1a037fdd


When user do not select Other option

https://github.com/google/site-kit-wp/assets/94359491/5bbb1e1c-b818-45e8-9977-adcd455d2559

When user edit questions

https://github.com/google/site-kit-wp/assets/94359491/2f84b6a5-c963-4bb7-b812-d7e9a0cdc9c3

</details>
